static/js/typeahead_helper.js

```diff
@@ -144,12 +169,54 @@ exports.compare_by_pms = function (user_a, user_b) {
     return 1;
 };
 
-exports.sort_for_at_mentioning = function (objs, current_stream) {
+exports.sort_for_at_mentioning = function (objs, current_stream, current_subject) {
     var objs_split = split_by_subscribers(objs, current_stream);
 
-    var subs_sorted = objs_split.subscribers.sort(exports.compare_by_pms);
-    var non_subs_sorted = objs_split.non_subscribers.sort(exports.compare_by_pms);
-    return subs_sorted.concat(non_subs_sorted);
+    var pmed_ids = [];
+    _.each(message_store.recent_private_messages || [], function (pm) {
+        pmed_ids.push.apply(pmed_ids, pm.user_ids_string.split(","));
+    });
+    var subbed_objs_split = split_by_pms(objs_split.subscribers, pmed_ids);
+    var non_subbed_objs_split = split_by_pms(objs_split.non_subscribers, pmed_ids);
+
+    var subs_pm_sorted = [];
+    var non_subs_pm_sorted = [];
+    var subs_no_pm_sorted = [];
+    var non_subs_no_pm_sorted = [];
```

Comment:
We use the strategy of splitting data into two arrays for sorting in a couple places in our code.  I think it's fine for totally simple cases, but it doesn't scale well to multiple splitting critieria.

This code is pretty hard to read.

I would suggest we instead use a strategy of having a compare function that triages sub-vs.non and pm-vs.non.

Here is an example:

https://github.com/zulip/zulip/blob/ee17f8fc8dcc8820dc73ab0774bb9c068fa44952/static/js/activity.js#L178